<a href="https://colab.research.google.com/github/NicolasCaminos/Python/blob/v0.1/Data_Science_Tutorials/API.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Estructura de archivos: Lectura

## Lectura de Archivos desde Api

In [1]:
import pandas as pd
# Importar la biblioteca 'requests' para hacer solicitudes HTTP
import requests

# Importar la biblioteca 'json' para trabajar con datos JSON
import json

# Definir la URL del servicio web que proporciona datos de seguimiento de COVID-19
url = 'https://api.covidtracking.com/v1/states/daily.json'

# Hacer una solicitud GET a la URL para obtener datos
response = requests.get(url)

# Imprimir la respuesta de la solicitud (esto mostrará información sobre la respuesta HTTP)
print(response)

# Verificar si la solicitud fue exitosa (código de estado HTTP 200)
# Caso contrario debería imprimir La solicitud no fue exitosa. Código de estado HTTP:', (código de estado HTTP: 404)
if response.status_code == 200:
    # Parsear la respuesta JSON y almacenarla en la variable 'data'
    data = response.json()

    # Ahora puedes trabajar con los datos en el formato JSON almacenados en 'data'
else:
    print('La solicitud no fue exitosa. Código de estado HTTP:', response.status_code)

<Response [200]>


In [2]:
#Vemos datos de la API
print(len(response.text))
texto=response.text

29061568


In [3]:
# Presentamos los datos en una tabla
jsondata=json.loads(texto)
df2=pd.DataFrame.from_dict(jsondata)
df2.head(8)

date state   positive  probableCases   negative  pending  \
0  20210307    AK    56886.0            NaN        NaN      NaN   
1  20210307    AL   499819.0       107742.0  1931711.0      NaN   
2  20210307    AR   324818.0        69092.0  2480716.0      NaN   
3  20210307    AS        0.0            NaN     2140.0      NaN   
4  20210307    AZ   826454.0        56519.0  3073010.0      NaN   
5  20210307    CA  3501394.0            NaN        NaN      NaN   
6  20210307    CO   436602.0        24786.0  2199458.0      NaN   
7  20210307    CT   285330.0        19621.0        NaN      NaN   

     totalTestResultsSource  totalTestResults  hospitalizedCurrently  \
0           totalTestsViral         1731628.0                   33.0   
1     totalTestsPeopleViral         2323788.0                  494.0   
2           totalTestsViral         2736442.0                  335.0   
3           totalTestsViral            2140.0                    NaN   
4           totalTestsViral         7908105.0                  963.0   
5           totalTestsViral        49646014.0                 4291.0   
6  totalTestEncountersViral         6415123.0                  326.0   
7           totalTestsViral         6520366.0                  428.0   

   hospitalizedCumulative  ...  dataQualityGrade  deathIncrease  \
0                  1293.0  ...              None              0   
1                 45976.0  ...              None             -1   
2                 14926.0  ...              None             22   
3                     NaN  ...              None              0   
4                 57907.0  ...              None              5   
5                     NaN  ...              None            258   
6                 23904.0  ...              None              3   
7                     NaN  ...              None              0   

   hospitalizedIncrease                                      hash  \
0                     0  dc4bccd4bb885349d7e94d6fed058e285d4be164   
1                     0  997207b430824ea40b8eb8506c19a93e07bc972e   
2                    11  50921aeefba3e30d31623aa495b47fb2ecc72fae   
3                     0  f77912d0b80d579fbb6202fa1a90554fc4dc1443   
4                    44  0437a7a96f4471666f775e63e86923eb5cbd8cdf   
5                     0  63c5c0fd2daef2fb65150e9db486de98ed3f7b72   
6                    18  444746cda3a596f183f3fa3269c8cab68704e819   
7                     0  38f39494134d22cd0d06e4ce63aa4a4ac10aa930   

   commercialScore negativeRegularScore negativeScore positiveScore  score  \
0                0                    0             0             0      0   
1                0                    0             0             0      0   
2                0                    0             0             0      0   
3                0                    0             0             0      0   
4                0                    0             0             0      0   
5                0                    0             0             0      0   
6                0                    0             0             0      0   
7                0                    0             0             0      0   

   grade  
0         
1         
2         
3         
4         
5         
6         
7         

[8 rows x 56 columns]

## Lectura desde Github

In [7]:
import pandas as pd

# Defino la URL del archivo CSV en GitHub
csv_url = 'https://raw.githubusercontent.com/NicolasCaminos/Python/master/Data_Science_Tutorials/fifa_eda_stats.csv'

try:
    # Cargo el archivo CSV en un DataFrame de pandas
    df = pd.read_csv(csv_url)

    # Ahora podes trabajar con los datos en el DataFrame 'df'
    print(df.head(8))  # Visualizo las primeras filas del DataFrame

except pd.errors.EmptyDataError:
    print('El archivo CSV está vacío o no se pudo cargar.')

except Exception as e:
    print('Ocurrió un error al cargar el archivo CSV:', e)

       ID               Name  Age Nationality  Overall  Potential  \
0  158023           L. Messi   31   Argentina       94         94   
1   20801  Cristiano Ronaldo   33    Portugal       94         94   
2  190871          Neymar Jr   26      Brazil       92         93   
3  193080             De Gea   27       Spain       91         93   
4  192985       K. De Bruyne   27     Belgium       91         92   
5  183277          E. Hazard   27     Belgium       91         91   
6  177003          L. Modrić   32     Croatia       91         91   
7  176580          L. Suárez   31     Uruguay       91         91   

                  Club    Value   Wage Preferred Foot  ...  Composure  \
0         FC Barcelona  €110.5M  €565K           Left  ...       96.0   
1             Juventus     €77M  €405K          Right  ...       95.0   
2  Paris Saint-Germain  €118.5M  €290K          Right  ...       94.0   
3    Manchester United     €72M  €260K          Right  ...       68.0   
4      Manche

## Glosario

1. API: Una API, o Interfaz de Programación de Aplicaciones por sus siglas en inglés (Application Programming Interface), es un conjunto de reglas y definiciones que permite que diferentes aplicaciones o software se comuniquen entre sí. En otras palabras, una API define cómo deben interactuar los componentes de software.

Al realizar una solicitud a una API mediante HTTP, el servidor generalmente responde con un código de estado HTTP que indica el resultado de la solicitud. Aquí hay algunos códigos de respuesta comunes que puedes encontrar al consultar una API:

- **200 OK:** La solicitud fue exitosa. La respuesta incluirá los datos solicitados en el cuerpo del mensaje.

- **201 Created:** La solicitud ha tenido éxito y se ha creado un nuevo recurso como resultado.

- **204 No Content:** La solicitud se ha completado con éxito, pero no hay contenido para enviar en la respuesta.

- **400 Bad Request:** La solicitud no pudo ser entendida o procesada por el servidor debido a una sintaxis incorrecta, parámetros inválidos u otros errores del cliente.

- **401 Unauthorized:** El cliente no está autorizado para acceder al recurso solicitado. Se requiere autenticación.

- **403 Forbidden:** El cliente está autenticado, pero no tiene permisos suficientes para acceder al recurso.

- **404 Not Found:** El recurso solicitado no se pudo encontrar en el servidor.

- **405 Method Not Allowed:** El método de solicitud utilizado (GET, POST, etc.) no está permitido para el recurso solicitado.

- **500 Internal Server Error:** Hubo un error interno en el servidor que impidió que se completara la solicitud.


2. **Request:** La palabra "request" (petición en español) se utiliza comúnmente en el contexto de las comunicaciones en línea y la programación para referirse a una solicitud de información o acción entre un cliente y un servidor. En el contexto de la web y el protocolo HTTP, una "request" o petición es un mensaje enviado por un cliente a un servidor para solicitar algún recurso o realizar alguna acción.

Existen varios tipos de peticiones en el protocolo HTTP, siendo los más comunes:

- **GET: Utilizado para recuperar datos del servidor. Se incluyen los parámetros en la URL. Es la que usaremos para traer los datos. **

- **POST:** Utilizado para enviar datos al servidor, a menudo en el cuerpo del mensaje HTTP. Se utiliza para enviar información que puede no ser adecuada para incluir en la URL, como datos de formularios.

- **PUT:** Utilizado para enviar datos al servidor para crear o actualizar un recurso en particular.

- **DELETE:** Utilizado para solicitar al servidor que elimine un recurso.

- **HEAD:** Similar a GET, pero se utiliza para obtener solo los encabezados de la respuesta, sin el cuerpo del mensaje.